Issue 134 read age class mixing matrix

WPettersson opened the pull request:
Adds MixingMatrix class, and tests

Closes: SCRC-134
Signed-off-by: William Pettersson <william.pettersson@glasgow.ac.uk>

magicicada commented:
Looks fine to me, but might be worth a peek from @aflag in case there's any testing interference or file placement implications?

aflag commented on simple_network_sim/mixing_matrix.py:
```diff
@@ -0,0 +1,132 @@
+            self._lower = int(a)
+            self._upper = int(b)
+
+    def contains(self, age):
```
Maybe define `__contains__` and use the `in` operator below?

WPettersson replied:
Done.

aflag commented on simple_network_sim/mixing_matrix.py:
```diff
@@ -0,0 +1,132 @@
+        self._matrix = {}
+        with open(infile, "r") as inp:
+            reader = csv.reader(inp)
+            headers = [AgeRange(text) for text in next(reader, None)[1:]]
```
No reason to pass `None`, since this will fail anyway if reader is empty

WPettersson replied:
Done.

aflag commented on simple_network_sim/mixing_matrix.py:
```diff
@@ -0,0 +1,132 @@
+        return not self == other
+
+    def __hash__(self):
+        return 1000*self._lower + self._upper  # We just need any hashing method
```
Maybe `hash((self._lower, self._upper))` would be better here?

WPettersson replied:
Done.

aflag commented on simple_network_sim/mixing_matrix.py:
```diff
@@ -0,0 +1,132 @@
+        if b is None:
+            if isinstance(a, tuple) and len(a) == 2:
+                self._lower = a[0]
+                self._upper = a[1]
```
When are we expecting to use `AgeRange(tuple)` or `AgeRange(a=int, b=int)`?

WPettersson replied:
The tuple was given as an example in the original issue, as `returnDict[(0,5)][(5,18)]`. The two-int version I don't see a current need for at the moment, but felt trivial enough to include that it was worth writing up.

aflag commented on tests/unit/test_mixing_matrix.py:
```diff
@@ -0,0 +1,33 @@
+def test_sampleMixingMatrix(mixing_matrix):
+    mm = MixingMatrix(mixing_matrix)
+    assert mm[28][57] == 0.097096431
+    assert mm["[30,40)"][75] == 0.026352071
```
I think this interface may be a bit confusing. If I tried to query mm["[30,35)"][75] I'd get an error, right?

WPettersson replied:
With the current implementation and reading the sample mixing matrix, it would raise a KeyError on "[30,35)" yes. I wasn't sure if there was a need to allow alternate age ranges (i.e. ones not defined in the mixing matrix) because it might imply higher accuracy data than what we actually have (i.e., someone might try running with age ranges [x, x+1) for each age x, which would "run" fine but wouldn't be using any more accurate data for the mixing matrix).

aflag replied:
I see, I think you're right. "[30,40)" is the age class name, shouldn't really be thought of as a range.

aflag commented on simple_network_sim/mixing_matrix.py:
```diff
@@ -0,0 +1,132 @@
+
+    """
+
+    def __init__(self, infile):
```
Maybe it'd be interesting to raise an exception or print an error if there are any overlapping ranges. If someone messes up one of the ranges it could cause hard to detect bugs otherwise.

WPettersson replied:
Added a function to detect overlapping ranges, and adding some tests for these too.

aflag commented:
Looks good. I had some small comments. Do you think it would make sense to add this code to the `loaders` module? I think it servers the same sort of purpose, right?

WPettersson commented:
Sure, definitely makes sense. I'll add a commit for that too.

WPettersson commented:
Okay I've added some more testing, and pushed it into the loaders module. I wasn't sure exactly how you wanted the loaders module laid out, so I added it all to the end of the file, feel free to let me know if that isn't quite right.

bobturneruk commented:
Suggest merge on @aflag approval.

aflag commented on simple_network_sim/loaders.py:
```diff
@@ -115,3 +117,172 @@ def readNodeAttributesJSON(filename):
+            # Check for any overlap in the column headers
+            for one in headers:
+                for two in headers:
+                    if one == two:
```
minor thing, we could use enumerate(headers) here and compare the headers position rather than value. This way we complain if there are repeated lines too. Eg.:

```
for i,one in enumerate(headers):
    for j,two in enumerate(headers):
        if i != j:
            _check_overlap(one, two)
```

WPettersson replied:
Used enumerate to check for equality. Also removed the lines two above, which checked for duplicate entries since that was now redundant.

aflag commented on tests/unit/test_loaders.py:
```diff
@@ -248,3 +248,100 @@ def test_genGraphFromContactFile(commute_moves):
+        fp.flush()
+        with pytest.raises(Exception) as e_info:
+            loaders.MixingMatrix(fp.name)
+        assert e_info.value.args[0] == "Overlap in age ranges with [0,15) and [10,30)"
```
We have different test cases here. I think that, if we split into multiple test functions, it will be easier to see what went wrong when a test fails

- [ ] split into multiple test functions

WPettersson replied:
I split this function into two test cases - one for overlapping ranges and one for duplicate headers.

aflag commented:
I had some minor comments, I'll leave it approved (I don't know how github handle tasks, but in my day to day job we usually leave tasks open and approve so the PR creator can just merge after they deal with the tasks).

bobturneruk commented:
@WPettersson can you merge when you've done the stuff @aflag was after, please? Thanks for your efforts doing and reviewing this!
